docs/csharp/programming-guide/classes-and-structs/ref-returns.md

```diff
@@ -0,0 +1,94 @@
+
+Reference return values can produce more compact code, as well as allow an object to expose only the individual data items, such as an array element, that are of interest to the caller. This reduces the likelihood that the caller will inadvertently modify the object's state.
+
+There are some restrictions on the value that a method can return as a reference return value. These include:
+
+- The return value cannot be `void`. Attempting to define a method with a `void` reference return value generates compiler error CS1547, "Keyword 'void' cannot be used in this context."
+ 
+- The return value cannot be a local variable in the procedure that returns it; it must have global scope. Attempting to return a local variable generates compiler error CS8168, "Cannot return local 'obj' by reference because it is not a ref local."
+
+- The return value cannot be a `null`. Attempting to return `null` generates compiler error CS8156, "An expression cannot be used in this context because it may not be returned by reference."
+
+   If a method with a ref return needs to return a null value, you can either return a null (uninstantiated) value for a reference type or a [nullable type](../nullable-types/index.md) for a value type.
+ 
+- The return value cannot be a constant, an enumeration member, or a property of a `class` or `struct`. Attempting to return these generates compiler error CS8156, "An expression cannot be used in this context because it may not be returned by reference."
+
+In addition, because an asynchronous method may return before it has finished execution and its return value is known, reference return values are not allowed on `async` methods.
+ 
+## Defining a ref return value
+
+You define a ref return value by adding the [ref](../../language-reference/keywords/ref.md) keyword to the return type of the method signature. For example, the following signature indicates that the `GetContactInformation` property returns a reference to a `Person` object to the caller:
+
+```csharp
+public ref Person GetContactInformation(string fname, string lname);
+```
+
+In addition, the name of the object returned by each [return](../../language-reference/keywords/return.md) statement in the method body must be preceded by the [ref](../../language-reference/keywords/ref.md) keyword. For example, the following `return` statement returns a `Person` object named `p` by reference:
+
+```csharp
+return ref p;
+```
+
+## Consuming a ref return value
+
+A caller can handle a ref return value in either of two ways:
+
+- As an ordinary value returned by value from a method. The caller can choose to ignore that the return value is a reference return value. In this case, any changes made to the value returned by the method call are not reflected in the state of the called type.
```

Comment:
You should also add that if the type is a value type, the value is copied to new storage. (That's why changes made are not reflected in the called type).